### PR TITLE
VideoPress: only enqueue the bridge script when the player is rendered

### LINKED ISSUE #24765: VideoPress: avoid enqueuing js on pages that do not need it.
This is a follow-up to #23949. We now enqueue `modules/videopress/js/videopress-token-bridge.js` on all pages of websites where the module is enabled:
https://github.com/Automattic/jetpack/blob/304e134beec8ec1d3dd79e04288bc14a5a849c6c/projects/plugins/jetpack/modules/videopress/class.jetpack-videopress.php#L72-L95

That seems problematic since we do not need it on all those pages.

@Automattic/greenhouse Do you think we could be a bit more restrictive there?

### PR DESCRIPTION
**Pull request by jgcaruso**
Fixes #24765

#### Changes proposed in this Pull Request:

* Moved front end bridge script enqueueing to VideoPress player render functions
* removed duplicate enqueue statement for admin pages (different version number was causing double enqueue)

#### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?

#### Does this pull request change what data or activity we track or use?

Nope
#### Testing instructions:

* Use a site with VideoPress enabled
* Create a post with a VideoPress shortcode
* Create a post with a `/video` block
* Create a post and paste a VideoPress url in the page (uses oembed)
* Create a post with no VideoPress content in it
* View the pages in the front end, the `videopress-token-bridge.js` script should be rendered only on pages with VideoPress content
* View the homepage, it should also not load the bridge script unless your post summaries include the VideoPress player
* Make one or more of the videos on your posts private
* View the post on the front end to ensure the bridge is working correctly
* Visit `/wp-admin`, private videos in the post editor and the Media Library should still be viewable

### CODE REVIEW

**Review by roundhill — COMMENTED** [review]
Testing really well, just needs conflicts fixed for ✅

**Review by roundhill — DISMISSED** [review]
👍

**Comment by sdixon194** [comment]
Looks like it doesn't like the unused `$attr` attribute in `video_enqueue_bridge_when_oembed_present`. If it's not needed, can it be removed, or otherwise throw a `phpcs:ignore VariableAnalysis.CodeAnalysis.VariableAnalysis.UnusedVariable` on there?

**Comment by jgcaruso** [comment]
> Looks like it doesn't like the unused `$attr` attribute in `video_enqueue_bridge_when_oembed_present`. If it's not needed, can it be removed, or otherwise throw a `phpcs:ignore VariableAnalysis.CodeAnalysis.VariableAnalysis.UnusedVariable` on there?

I think I'll put the ignore on there... those are part of the filter's signature so I'd rather not mess with it, unless that is a normal thing to do?

**Comment by sdixon194** [comment]
> those are part of the filter's signature so I'd rather not mess with it, unless that is a normal thing to do?

Yup, that makes sense and works for me!

**Review by samiff — DISMISSED** [review]

**Review by jeherve — DISMISSED** [review]
Looks like the tests are failing now:

```
PHP Fatal error:  Class 'VideoPress_Scheduler' not found in /tmp/wordpress-latest/src/wp-content/plugins/jetpack/modules/videopress/class.jetpack-videopress.php on line 71
  ..............................
  Fatal error: Class 'VideoPress_Scheduler' not found in /tmp/wordpress-latest/src/wp-content/plugins/jetpack/modules/videopress/class.jetpack-videopress.php on line 71
```

**Comment by jgcaruso** [comment]
Rebased and updated the files since a refactor in another PR moved around a bunch of the files.

Also noticed an issue with Private videos no longer loading as a result of that refactor, which is fixed separately in https://github.com/Automattic/jetpack/pull/25638

**Review by samiff — DISMISSED** [review]

**Comment by leogermani** [comment]
@jgcaruso One question related to this. I was under the impression that the class VideoPress_Player was only used by the shortcode. Is that true? Or is it only used by the block in some way?

Also, I thought the jwt token bridge was used by all player instances, either block or shortcodes. Is it?

For context, we are cherry picking what will be migrated to the package (to be shared with the stand-alone plugin) and what will not. See pe4Cmx-40-p2

**Comment by jgcaruso** [comment]
> I was under the impression that the class VideoPress_Player was only used by the shortcode. Is that true? Or is it only used by the block in some way?

The code for the VideoPress_Player does appear to only be used by the shortcode. I haven't noticed any other references to it. Now that you call this out, I'm wondering why I put the `embed_oembed_html` filter handler in the `shortcode.php` file. It might make more sense somewhere else.

Given you're reorganizing all of this code, where would you advise me to put the `embed_oembed_html` filter handler so that it gets included where it needs to for both the standalone plugin and Jetpack? This is the handler that will enqueue the jwt token bridge for the block when present on a page.

> Also, I thought the jwt token bridge was used by all player instances, either block or shortcodes. Is it?

The jwt token bridge is used by all player instances. More specifically, it is used directly by the videopress iframe that gets rendered by them in order for private videos to load correctly.

**Comment by leogermani** [comment]
> I'm wondering why I put the embed_oembed_html filter handler in the shortcode.php file. It might make more sense somewhere else.

The custom oembed regexp registration used to live there and maybe you had this on your memory.

This has been already moved to the package:

https://github.com/Automattic/jetpack/blob/trunk/projects/packages/videopress/src/class-initializer.php#L72

So I guess this filter should also be added there in the initializer.

> The jwt token bridge is used by all player instances. More specifically, it is used directly by the videopress iframe that gets rendered by them in order for private videos to load correctly.

Gotcha, and it's added via the oembed filter. So it definitely needs to go in the package.

**Comment by jgcaruso** [comment]
@leogermani thanks for that! I've moved the oembed filter in with the oembed initializers so it should be good to go now.

File: projects/packages/videopress/src/class-initializer.php
Comment (leogermani) [inline projects/packages/videopress/src/class-initializer.php]: nitpick: can we use `__CLASS__` here?

**Review by leogermani — DISMISSED** [review]
Tests well!

Left a minor comment

File: projects/packages/videopress/src/class-initializer.php
Comment (leogermani) [inline projects/packages/videopress/src/class-initializer.php]: Don`t we need to add support to `v.wordpress.com` as well?

Comment (jgcaruso) [reply projects/packages/videopress/src/class-initializer.php]: I don't think I've ever seen `v.wordpress.com` actually used anywhere as an embed, and that url doesn't seem to work for viewing videos on its own. Its possible it was an old way of accessing videos.

I suppose I could add it for completeness.

Comment (jgcaruso) [reply projects/packages/videopress/src/class-initializer.php]: I added a 2nd condition for the `https://v.wordpress.com/GUID` embeds. Since the regex just translates them into a standard embed it was easy to test as well, even though the url itself doesn't do anything.

**Review by leogermani — DISMISSED** [review]

**Review by jeherve — COMMENTED** [review]
I believe this will need one last update with trunk before we can merge it.

File: projects/plugins/jetpack/changelog/update-videopress-only-enqueue-bridge-when-player-is-rendered
Comment (jeherve) [inline projects/plugins/jetpack/changelog/update-videopress-only-enqueue-bridge-when-player-is-rendered]: I would suggest making this a bit more end-user friendly. Something like:

```suggestion
VideoPress: ensure that the player's resources are only loaded when necessary, when a video was added to the page.
```

File: projects/packages/videopress/src/class-initializer.php
Comment (jeherve) [inline projects/packages/videopress/src/class-initializer.php]: Should we be a bit more explicit, like below, to avoid matching for things like `https://mysite.com/https://videopressscom/v/AxWYhwqp`:

```suggestion
		if ( preg_match( '/^https?:\/\/(video\.wordpress\.com|videopress\.com)\/(v|embed)\//', $url )
```

Maybe you could add a test for this, too, to ensure we catch the right things?

**Review by jeherve — CHANGES_REQUESTED** [review]

File: projects/packages/videopress/src/class-initializer.php
Comment (jgcaruso) [reply projects/packages/videopress/src/class-initializer.php]: Thanks for catching that... I just used the existing regex from earlier in the file and didn't consider that it could have been incorrect.

**Comment by jgcaruso** [comment]
@jeherve I've made the changes and added a test for the function that ensures the proper video url types trigger the enqueue function.

There are a couple of github checks that are failing that I'm not sure how to address, unless they're ok?

One is the `Linting / Lock files are up to date (pull_request)` and the other is a file sync check, but it might be easier to clean that one up after the fact since that code has changed a bit and should be cleaner once this PR goes in. Based on whats missing (looks like new code if I'm reading the file sync checker tool correctly for `projects/plugins/jetpack/modules/videopress/class.videopress-gutenberg.php`), I suspect it was skipped previously as well.

**Review by jeherve — CHANGES_REQUESTED** [review]
Thanks for adding the tests!

> One is the Linting / Lock files are up to date (pull_request) and the other is a file sync check, but it might be easier to clean that one up after the fact since that code has changed a bit and should be cleaner once this PR goes in. Based on whats missing (looks like new code if I'm reading the file sync checker tool correctly for projects/plugins/jetpack/modules/videopress/class.videopress-gutenberg.php), I suspect it was skipped previously as well.

You should be able to fix the linting / lock file issues by updating with latest trunk once again, and running the commands suggested by the failure messages in this view:
https://github.com/Automattic/jetpack/pull/24985/files
(since you made changes to a composer.json file, you have to update the matching composer.lock files, as well as lock files for projects that have the VideoPress package as a dependency).

The new test seems to be failing though:
```PHP Fatal error:  Declaration of Automattic\Jetpack\InitializerTest::setUp() must be compatible with PHPUnit\Framework\TestCase::setUp(): void in /home/runner/work/jetpack/jetpack/projects/packages/videopress/tests/php/test-class-initializer.php on line 18
  
  Fatal error: Declaration of Automattic\Jetpack\InitializerTest::setUp() must be compatible with PHPUnit\Framework\TestCase::setUp(): void in /home/runner/work/jetpack/jetpack/projects/packages/videopress/tests/php/test-class-initializer.php on line 18
```

**Comment by jgcaruso** [comment]
@jeherve Ok I was able to fix up the composer files using the `fixup` tool and tests are working again after I looked at some of the other tests to see how they were getting around the php 5.6 lack of support for `:void`.

The file sync issue also looks like its gone after the deploy of a diff that was on hold for a while.

This PR **_should_** be good to go now 🤞

**Review by jeherve — APPROVED** [review]

**Comment by jgcaruso** [comment]
deployed r250884-wpcom

**Pull request merged**
